compiler/rustc_type_ir/src/fast_reject.rs

```diff
@@ -234,17 +236,32 @@ impl<I: Interner, const INSTANTIATE_LHS_WITH_INFER: bool, const INSTANTIATE_RHS_
         })
     }
 
-    pub fn types_may_unify(self, lhs: I::Ty, rhs: I::Ty) -> bool {
+    pub fn types_may_unify(&mut self, lhs: I::Ty, rhs: I::Ty) -> bool {
+        if self.cache.contains(&(lhs, rhs)) {
+            return true;
+        }
```

Comment:
you can def move this cache access (and inserts) out of the rhs check.

if `types_may_unify` returns right away we won't need to cache, only need to cache for branches which may deeply recur